File: java/rocks/sakira/sakurarosea/common/block/SakuraBarrelBlock.java
package rocks.sakira.sakurarosea.common.block;

import net.minecraft.block.BarrelBlock;
import net.minecraft.block.Block;
import net.minecraft.block.BlockState;
import net.minecraft.block.Blocks;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.ItemStack;
import net.minecraft.stats.Stats;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.tileentity.TileEntityType;
import net.minecraft.util.ActionResultType;
import net.minecraft.util.Hand;
import net.minecraft.util.LazyValue;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.BlockRayTraceResult;
import net.minecraft.world.IBlockReader;
import net.minecraft.world.World;
import net.minecraft.world.server.ServerWorld;
import rocks.sakira.sakurarosea.common.tileentities.SakuraBarrelTileEntity;
import rocks.sakira.sakurarosea.common.tileentities.TileEntities;

import javax.annotation.Nullable;
import java.util.Random;

public class SakuraBarrelBlock extends BarrelBlock {
    private final LazyValue<TileEntityType<SakuraBarrelTileEntity>> tileEntityType;

    public SakuraBarrelBlock() {
        super(Block.Properties.from(Blocks.BARREL));

        this.tileEntityType = new LazyValue<>(TileEntities.SAKURA_BARREL_ENTITY);
    }

    @Override
    public ActionResultType onBlockActivated(BlockState state, World world, BlockPos blockPos, PlayerEntity playerEntity, Hand handIn, BlockRayTraceResult hit) {
        if (!world.isRemote) {
            TileEntity tileEntity = world.getTileEntity(blockPos);

            if (tileEntity instanceof SakuraBarrelTileEntity) {
                playerEntity.openContainer((SakuraBarrelTileEntity) tileEntity);
                playerEntity.addStat(Stats.OPEN_BARREL);
            }
        }

        return ActionResultType.SUCCESS;
    }

    public TileEntityType<SakuraBarrelTileEntity> getTileEntityType() {
        return this.tileEntityType.getValue();
    }

    @Nullable
    @Override
    public TileEntity createNewTileEntity(IBlockReader worldIn) {
        return this.getTileEntityType().create();
    }

    @Override
    public void tick(BlockState state, ServerWorld world, BlockPos blockPos, Random rand) {
        TileEntity tileEntity = world.getTileEntity(blockPos);
        if (tileEntity instanceof SakuraBarrelTileEntity) {
            ((SakuraBarrelTileEntity) tileEntity).tick();
        }
    }

    @Override
    public void onBlockPlacedBy(World world, BlockPos blockPos, BlockState state, @Nullable LivingEntity placer, ItemStack itemStack) {
        if (itemStack.hasDisplayName()) {
            TileEntity tileEntity = world.getTileEntity(blockPos);
            if (tileEntity instanceof SakuraBarrelTileEntity) {
                ((SakuraBarrelTileEntity) tileEntity).setCustomName(itemStack.getDisplayName());
            }
        }
    }
}

File: src/main/kotlin/io/github/faecraft/sakurarosea/block/SakuraBarrelBlock.kt
package io.github.faecraft.sakurarosea.block

import net.minecraft.block.BarrelBlock
import net.minecraft.block.BlockState
import net.minecraft.block.entity.BlockEntity
import net.minecraft.entity.LivingEntity
import net.minecraft.entity.mob.PiglinBrain
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.ItemStack
import net.minecraft.server.world.ServerWorld
import net.minecraft.stat.Stats
import net.minecraft.util.ActionResult
import net.minecraft.util.Hand
import net.minecraft.util.hit.BlockHitResult
import net.minecraft.util.math.BlockPos
import net.minecraft.world.BlockView
import net.minecraft.world.World
import io.github.faecraft.sakurarosea.blockentity.SakuraBarrelBlockEntity
import java.util.*

class SakuraBarrelBlock(settings: Settings) : BarrelBlock(settings) {
    override fun onUse(
        state: BlockState,
        world: World,
        pos: BlockPos,
        player: PlayerEntity,
        hand: Hand,
        hit: BlockHitResult
    ): ActionResult {
        return if (world.isClient) {
            ActionResult.SUCCESS
        } else {
            val blockEntity = world.getBlockEntity(pos)

            if (blockEntity is SakuraBarrelBlockEntity) {
                player.openHandledScreen(blockEntity)
                player.incrementStat(Stats.OPEN_BARREL)

                PiglinBrain.onGuardedBlockInteracted(player, true)
            }

            ActionResult.SUCCESS
        }
    }

    override fun scheduledTick(state: BlockState, world: ServerWorld, pos: BlockPos, random: Random) {
        val blockEntity = world.getBlockEntity(pos)

        if (blockEntity is SakuraBarrelBlockEntity) {
            blockEntity.tick()
        }
    }

    override fun createBlockEntity(world: BlockView): BlockEntity? {
        return SakuraBarrelBlockEntity()
    }

    override fun onPlaced(
        world: World,
        pos: BlockPos,
        state: BlockState,
        placer: LivingEntity?,
        itemStack: ItemStack
    ) {
        if (itemStack.hasCustomName()) {
            val blockEntity = world.getBlockEntity(pos)

            if (blockEntity is SakuraBarrelBlockEntity) {
                blockEntity.customName = itemStack.name
            }
        }
    }
}
